File: content/tracks/dsa-javascript/concepts/08-linked-list-patterns/linked-list-patterns_test.js
const assert = require("assert");
const { ListNode, listFromArray, listToArray, reverseList, hasCycle, mergeTwoSortedLists } = require("./stub");

let passed = 0, failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e) {
    console.log(`  FAIL: ${name} — ${e.message}`);
    failed++;
  }
}

// reverseList
test("reverseList: standard list", () => {
  assert.deepStrictEqual(listToArray(reverseList(listFromArray([1, 2, 3, 4, 5]))), [5, 4, 3, 2, 1]);
});
test("reverseList: single node", () => {
  assert.deepStrictEqual(listToArray(reverseList(listFromArray([1]))), [1]);
});
test("reverseList: null input", () => {
  assert.strictEqual(reverseList(null), null);
});
test("reverseList: two nodes", () => {
  assert.deepStrictEqual(listToArray(reverseList(listFromArray([1, 2]))), [2, 1]);
});
test("reverseList: new head is old tail", () => {
  const newHead = reverseList(listFromArray([1, 2, 3]));
  assert.strictEqual(newHead.val, 3);
});

// hasCycle
test("hasCycle: no cycle", () => {
  assert.strictEqual(hasCycle(listFromArray([1, 2, 3, 4])), false);
});
test("hasCycle: null input", () => {
  assert.strictEqual(hasCycle(null), false);
});
test("hasCycle: single node no cycle", () => {
  assert.strictEqual(hasCycle(listFromArray([1])), false);
});
test("hasCycle: cycle back to head", () => {
  const n1 = new ListNode(1);
  const n2 = new ListNode(2);
  const n3 = new ListNode(3);
  n1.next = n2; n2.next = n3; n3.next = n1;
  assert.strictEqual(hasCycle(n1), true);
});
test("hasCycle: cycle to middle node", () => {
  const n1 = new ListNode(1);
  const n2 = new ListNode(2);
  const n3 = new ListNode(3);
  const n4 = new ListNode(4);
  n1.next = n2; n2.next = n3; n3.next = n4; n4.next = n2;
  assert.strictEqual(hasCycle(n1), true);
});

// mergeTwoSortedLists
test("mergeTwoSortedLists: standard merge", () => {
  assert.deepStrictEqual(
    listToArray(mergeTwoSortedLists(listFromArray([1, 2, 4]), listFromArray([1, 3, 4]))),
    [1, 1, 2, 3, 4, 4]
  );
});
test("mergeTwoSortedLists: both empty", () => {
  assert.strictEqual(mergeTwoSortedLists(null, null), null);
});
test("mergeTwoSortedLists: first empty", () => {
  assert.deepStrictEqual(listToArray(mergeTwoSortedLists(null, listFromArray([1, 3]))), [1, 3]);
});
test("mergeTwoSortedLists: second empty", () => {
  assert.deepStrictEqual(listToArray(mergeTwoSortedLists(listFromArray([2, 4]), null)), [2, 4]);
});
test("mergeTwoSortedLists: equal elements", () => {
  assert.deepStrictEqual(
    listToArray(mergeTwoSortedLists(listFromArray([1, 1]), listFromArray([1, 1]))),
    [1, 1, 1, 1]
  );
});
test("mergeTwoSortedLists: single elements", () => {
  assert.deepStrictEqual(
    listToArray(mergeTwoSortedLists(listFromArray([5]), listFromArray([2]))),
    [2, 5]
  );
});
test("mergeTwoSortedLists: disjoint ranges", () => {
  assert.deepStrictEqual(
    listToArray(mergeTwoSortedLists(listFromArray([1, 2, 3]), listFromArray([4, 5, 6]))),
    [1, 2, 3, 4, 5, 6]
  );
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/dsa-python/concepts/08-linked-list-patterns/linked-list-patterns_test.py
import sys
from stub import ListNode, reverse_list, has_cycle, merge_two_sorted_lists

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


def make_list(vals: list[int]) -> "ListNode | None":
    if not vals:
        return None
    head = ListNode(vals[0])
    curr = head
    for v in vals[1:]:
        curr.next = ListNode(v)
        curr = curr.next
    return head


def to_vals(head: "ListNode | None", limit: int = 200) -> list[int]:
    vals: list[int] = []
    curr = head
    seen: set[int] = set()
    while curr and id(curr) not in seen and len(vals) < limit:
        seen.add(id(curr))
        vals.append(curr.val)
        curr = curr.next
    return vals


# reverse_list
test("reverse: five elements", to_vals(reverse_list(make_list([1, 2, 3, 4, 5]))) == [5, 4, 3, 2, 1])
test("reverse: two elements", to_vals(reverse_list(make_list([1, 2]))) == [2, 1])
test("reverse: single element", to_vals(reverse_list(make_list([42]))) == [42])
test("reverse: None input", reverse_list(None) is None)
test("reverse: three elements", to_vals(reverse_list(make_list([1, 2, 3]))) == [3, 2, 1])

# has_cycle
test("cycle: no cycle (3 nodes)", has_cycle(make_list([1, 2, 3])) is False)
test("cycle: empty list", has_cycle(None) is False)
test("cycle: single node no cycle", has_cycle(ListNode(1)) is False)

n1 = ListNode(1)
n2 = ListNode(2)
n3 = ListNode(3)
n4 = ListNode(4)
n1.next = n2; n2.next = n3; n3.next = n4; n4.next = n2
test("cycle: cycle at node 2", has_cycle(n1) is True)

ns = ListNode(1)
ns.next = ns
test("cycle: self-loop", has_cycle(ns) is True)

# merge_two_sorted_lists
r = to_vals(merge_two_sorted_lists(make_list([1, 2, 4]), make_list([1, 3, 4])))
test("merge: standard", r == [1, 1, 2, 3, 4, 4], f"got {r}")
r = to_vals(merge_two_sorted_lists(None, make_list([0])))
test("merge: empty first list", r == [0], f"got {r}")
r = to_vals(merge_two_sorted_lists(make_list([0]), None))
test("merge: empty second list", r == [0], f"got {r}")
r = to_vals(merge_two_sorted_lists(None, None))
test("merge: both empty", r == [], f"got {r}")
r = to_vals(merge_two_sorted_lists(make_list([1, 3, 5]), make_list([2, 4, 6])))
test("merge: interleaved", r == [1, 2, 3, 4, 5, 6], f"got {r}")
r = to_vals(merge_two_sorted_lists(make_list([1, 2, 3]), make_list([4, 5, 6])))
test("merge: non-overlapping ranges", r == [1, 2, 3, 4, 5, 6], f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)
